### PR TITLE
[dataset] remove 'GenerateLocal'

### PR DESCRIPTION
**Pull request by ColdDivinity**
This PR removes `ActiveDataset::GenerateLocal` and enforces the recommended method of forming Thread networks, as discussed in <https://github.com/openthread/openthread/issues/6011>.

This is on top of <https://github.com/openthread/openthread/pull/6119>, please just review the last commit.

Test fixing progress:
- [ ] OTNS: https://github.com/openthread/ot-ns/pull/113
- [X] expect
- [ ] toranj - Need to update `wpantund`
- [x] NCP
- [X] Cert
- [x] External commissioner
- [X] OTCI

Depends-On: openthread/ot-ns#113
Depends-On: openthread/pyspinel#110
Depends-On: openthread/ot-commissioner#173

### CODE REVIEW

File: tests/scripts/thread-cert/thread_cert.py [hunk 1]
```diff
@@ -235,6 +219,33 @@ def _setUp(self):
         self._inspector = debug.Inspector(self)
         self._collect_test_info_after_setup()
 
+    def _set_up_active_dataset(self, node, params, is_leader):
+        dataset = {
+            'master_key': binascii.hexlify(config.DEFAULT_MASTER_KEY).decode(),
+            'panid': params['panid'],
+        }
+
+        if is_leader:
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/thread_cert.py:228]: why do we need to know if the node is leader or not?

Can we setup the exact same active dataset on every node no matter it's a Leader node or not?

**Comment by abtink** [comment]
This is a rather big change in the behavior and impacts a lot of use-cases/tests.

I understand it's desired to move in this direction but I feel it may be too big of change to remove `GenerateLocal()` fully. 
My suggestion is to keep the behavior and add a build feature config around it (`OT_CONFIG`) first.

**Comment by simonlingoogle** [comment]
> This is a rather big change in the behavior and impacts a lot of use-cases/tests.
> 
> I understand it's desired to move in this direction but I feel it may be too big of change to remove `GenerateLocal()` fully.
> My suggestion is to keep the behavior and add a build feature config around it (`OT_CONFIG`) first.

I am for the idea of removing `GenerateLocal()`. However, I think we need to find a working way without adding `is_leader` to all test topologies.

File: tests/scripts/thread-cert/thread_cert.py (hunk 1 above)
Comment (Moandor-y) [reply tests/scripts/thread-cert/thread_cert.py]: Looks like it works, and is better. Updated.

**Comment by Moandor-y** [comment]
> This is a rather big change in the behavior and impacts a lot of use-cases/tests.
> 
> I understand it's desired to move in this direction but I feel it may be too big of change to remove `GenerateLocal()` fully.
> My suggestion is to keep the behavior and add a build feature config around it (`OT_CONFIG`) first.

I'm OK with adding the config flag if it's too big to fully change all tests.

BTW, is `wpantund` deprecated? Can we remove toranj tests if so?

**Comment by simonlingoogle** [comment]
> I'm OK with adding the config flag if it's too big to fully change all tests.

Actually I don't think it's a good idea to introduce another configuration just for this behavior change. Connectivity is so important 
 to OpenThread that we should make it work in a unified way. 

> BTW, is `wpantund` deprecated? Can we remove toranj tests if so?

I think we should continue to support wpantund to some extent since it's required by community.

**Comment by abtink** [comment]
>  BTW, is wpantund deprecated? Can we remove toranj tests if so?

We **absolutely should not** remove `toranj` from OpenThread CI.
- `toranj` is actually a test-framework for OpenThread, NCP interface, and `wpantund`  (not just `wpantund`).
-  It is actually covering a lot of functionality and behaviors in OpenThread itself (a lot of edge-cases and special behaviors are only covered by `toranj` test-cases).
- It's mostly using `wpantund` as the driver/interface to be able to run/implement the tests.

So unless all the same set of tests are added/replicated in another test frame-work and added as part of CI,  it should not be removed.


Regarding `wpantund` we (from Google side) probably won't be actively adding new features/functionality in `wpantund`, but there is no reason to stop the testing of existing, functioning well-tested (shipping) code.

-------


I think we need to consider what is the problem are we trying to solve here with this PR and what is its impact and what's the impact of removing this behavior? 

- In my view, this is rather big change and therefore risky.
- So I suggest that rather than removing this fully, we first add a build config feature around it and disable it in certain cases.
- This way, it's not completely removed and can be enabled back quickly.
- Also there is no  need to change and update all the tests and code at the same time.

**Comment by Moandor-y** [comment]
> I think we need to consider what is the problem are we trying to solve here with this PR and what is its impact and what's the impact of removing this behavior?
> 
> * In my view, this is rather big change and therefore risky.
> * So I suggest that rather than removing this fully, we first add a build config feature around it and disable it in certain cases.
> * This way, it's not completely removed and can be enabled back quickly.
> * Also there is no  need to change and update all the tests and code at the same time.

IMO most codes in this change are for test cases. These tests should work with both GenerateLocal/non-GenerateLocal versions and are not risky. The actual behavior change is only 2 lines (in `ActiveDataset::StartLeader` and `MleRouter::BecomeLeader`).

I think one way is to split them into 2 PRs, one to make tests compatible, the other to do the behavior change. If anything goes wrong we can revert easily. Thoughts?

**Comment by simonlingoogle** [comment]
> I think one way is to split them into 2 PRs, one to make tests compatible, the other to do the behavior change. If anything goes wrong we can revert easily. Thoughts?

I think it's a good idea to do the tests change first.

Then, we can remove the GenerateLocal() and add corresponding tests to verify that it works:
- Verify that devices without active dataset can attach to an existing Thread network
- Verify that devices without active dataset can not become Leader by itself

**Comment by jwhui** [comment]
Superseded by #9215
